completions/bash/multipass

```diff
@@ -21,25 +21,105 @@ _multipass_complete()
 
-        local instances=$( \eval $cmd | \grep -Ev '(\+--|Name)' | \cut -d',' -f 1 )
+        local retval=$( \eval $cmd | \grep -Ev '(\+--|Name)' | \cut -d',' -f 1 )
+
+        for instance in $retval; do
+            instances="${instances} ${instance}"
+        done
+    }
+
+    _multipass_snapshots()
+    {
+        local instance="${cur%%.*}"
+        local cmd="multipass info --snapshot-overview "${instance}" --format=csv 2>/dev/null"
+        local retval=$( \eval $cmd | \grep -Ev '(\+--|Instance)' | \cut -d',' -f 2 )
+
+        for snapshot in $retval; do
+            snapshots="${snapshots} ${instance}.${snapshot}"
+        done
+    }
+
+    _multipass_instances_and_snapshots()
+    {
+        if [[ "${cur}" =~ '.' ]]; then
+            _multipass_snapshots
+            _multipass_filter_array "snapshots"
+        fi
+
+        _multipass_instances
+        _multipass_filter_array "instances"
+    }
+
+    _multipass_filter_array()
+    {
+        local backup=()
+
+        if [[ $1 == "instances" ]]; then
+            backup=$instances
+            instances=""
+        elif [[ $1 == "snapshots" ]]; then
+            backup=$snapshots
+            snapshots=""
+        fi
 
         local found
+        _get_comp_words_by_ref -n := -w WORDS -i CWORD cur prev
+        for item in $backup; do
+            found=0
+            for ((i=2; i<CWORD; i++)); do
+                if [[ "${WORDS[i]}" == ${item} ]]; then
+                    found=1
+                    break
+                fi
+            done
+            if [ ${found} == 0 ]; then
+                if [[ $1 == "instances" ]]; then
+                    instances="${instances} ${item}"
+                elif [[ $1 == "snapshots" ]]; then
+                    snapshots="${snapshots} ${item}"
+                fi
+            fi
+        done
+    }
 
+    _multipass_filter_snapshots()
+    {
+        local snapshots_bak="${snapshots}"
+        snapshots=""
+        local found
         _get_comp_words_by_ref -n := -w WORDS -i CWORD cur prev
-        for instance in $instances; do
+        for snapshot in $snapshots_bak; do
             found=0
             for ((i=2; i<CWORD; i++)); do
-                if [[ "${WORDS[i]}" == ${instance} ]]; then
+                if [[ "${WORDS[i]}" == $snapshot ]]; then
                     found=1
                     break
                 fi
             done
             if [ ${found} == 0 ]; then
-                opts="${opts} ${instance}"
+                snapshots="${snapshots} ${snapshot}"
             fi
         done
     }
 
+    _multipass_instances_unique()
+    {
+        _multipass_instances $1
+        _multipass_filter_array "instances"
+    }
+
+    _multipass_restorable_instances()
+    {
+        local retval=$( multipass info --all --no-runtime-information --format=csv \
+                           | \awk -F'("[^"]*")?,|\r?\n' '$2 == "Stopped" && $16 > 0' \
```

Comment:
Why do you need such a complex field separator? Things in between quotes are entries, not separators themselves, right? Are you trying to gobble quoted commas? But when would they come about? Doesn't just `-F',|\r?\n'` work?

Reply:
Oh, I think this might have been because at one point I was expecting nested commas within a single field (eg. multiple ip addresses). Having `--no-runtime-information` makes this a non-issue though.